### PR TITLE
[sc-memory][tests][changelog] Edit scs parser according to standart

### PR DESCRIPTION
**Pull request by smousy**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation

### CODE REVIEW

File: sc-memory/sc-memory/scs/scs.g4 [hunk 1]
```diff
@@ -51,12 +51,17 @@ content returns [ElementHandle handle]
   : ('_' { $ctx->isVar = true; } )?
       c=CONTENT_BODY
     {
-      std::string const v = $ctx->c->getText();
+      std::string v = $ctx->c->getText();
       SC_ASSERT(v.size() > 1, ());
+      if(v.find('!')==0)
+      {
+        v = v.substr(1, v.size()-2);
+      }
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/scs/scs.g4:59]:
```suggestion
      if (v.find('!') == 0)
      {
        v = v.substr(1, v.size() - 2);
      }
```

File: sc-memory/sc-memory/scs/scs.g4 [hunk 2]
```diff
@@ -78,10 +83,10 @@ connector returns [std::string text]
     | '-|>' | '<|-' | '-/>' | '</-'
     | '~>' | '<~' | '~|>' | '<|~'
     | '~/>' | '</~' | '_<>' | '_>' | '_<'
-    | '_..>' | '_<..' | '_->' | '_<-'
-    | '_<=>' | '_=>' | '_<=' | '_-|>' | '_<|-'
-    | '_-/>' | '_</-' | '_~>' | '_<~'
-    | '_~|>' | '_<|~' | '_~/>' | '_</~' )
+    | '_..>' | '_<..' | '<.._' | '_->' | '_<-'| '<-_'
+    | '_<=>' | '_=>' | '_<=' |'<=_' | '_-|>' | '_<|-' | '<|-_'
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/scs/scs.g4:87]: `<=_ node` is equal to `<= _node` and `<=_node`

Comment (NikitaZotov) [reply sc-memory/sc-memory/scs/scs.g4:87]: All `x_` arcs have such behaviour

File: sc-memory/sc-memory/scs/scs_types.cpp [hunk 3]
```diff
@@ -22,30 +22,37 @@ TypeResolver::MapType TypeResolver::ms_connectorToType = {
     {"<=", ScType::EdgeDCommonConst},
     {"_=>", ScType::EdgeDCommonVar},
     {"_<=", ScType::EdgeDCommonVar},
+    {"<=_", ScType::EdgeDCommonVar},
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/scs/scs_types.cpp:25]: You should to to remove this features. They broke scs-texts handling

File: sc-memory/tests/scs/units/test_scs_common.cpp [hunk 4]
```diff
@@ -209,14 +209,16 @@ TEST(scs_common, links)
       "a -> \"file://data.txt\";;"
       "b -> [x];;"
       "c -> _[];;"
-      "d -> [];;";
+      "d -> [];;"
+      "e -> ![lala]!;;"
```
Comment (NikitaZotov) [inline sc-memory/tests/scs/units/test_scs_common.cpp:213]: Add, please, case with empty link class ![]!

File: sc-memory/tests/scs/units/test_scs_common.cpp [hunk 5]
```diff
@@ -288,6 +294,36 @@ TEST(scs_common, edges)
   }
 }
 
+TEST(scs_common, reversed_edges)
+{
+  std::string const data =
```
Comment (NikitaZotov) [inline sc-memory/tests/scs/units/test_scs_common.cpp:299]: This test can be ivalid. Recheck it, please

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: sc-memory/sc-memory/scs/scs.g4 (hunk 2 above)
Comment (smousy) [reply sc-memory/sc-memory/scs/scs.g4:87]: Add tests, it works correctly

File: sc-memory/sc-memory/scs/scs_types.cpp (hunk 3 above)
Comment (smousy) [reply sc-memory/sc-memory/scs/scs_types.cpp:25]: Add tests, it works correctly

File: sc-memory/tests/scs/units/test_scs_common.cpp (hunk 5 above)
Comment (smousy) [reply sc-memory/tests/scs/units/test_scs_common.cpp:299]: It works correctly

File: sc-memory/sc-memory/scs/scs.g4 (hunk 1 above)
Comment (NikitaZotov) [reply sc-memory/sc-memory/scs/scs.g4:59]: Not corrected

File: sc-memory/tests/scs/units/test_scs_common.cpp (hunk 4 above)
Comment (NikitaZotov) [reply sc-memory/tests/scs/units/test_scs_common.cpp:213]: Not corrected

**Review by NikitaZotov — APPROVED** [review]

**Review by KovalM — CHANGES_REQUESTED** [review]
In current implementation, ScsParser doesn't see any different between [] and ![]. In memory both would be loaded as ScType::LinkConst. 

Ideally, we should add ScType::LinkConstClass but current byte string with sc types doesn't have spaces for new label. We should create issue for support this type in memory.

Until fix this issue, we should mark this sc-typ as basic edge from keynode like sc_link_class

File: sc-memory/sc-memory/scs/scs.g4 [hunk 6]
```diff
@@ -51,9 +51,16 @@ content returns [ElementHandle handle]
   : ('_' { $ctx->isVar = true; } )?
       c=CONTENT_BODY
     {
-      std::string const v = $ctx->c->getText();
+      std::string v = $ctx->c->getText();
       SC_ASSERT(v.size() > 1, ());
-      $ctx->handle = m_parser->ProcessContent(v.substr(1, v.size() - 2), $ctx->isVar);
+      if (v.find('!') == 0)
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/scs/scs.g4:56]: Please, check it in ProcessContent, don't pass any other paramaters to this method

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

**Review by NikitaZotov — APPROVED** [review]

**Review by KovalM — APPROVED** [review]

**Pull request merged**